packages/eslint-plugin/src/rules/strict-void-return.ts

```diff
@@ -0,0 +1,636 @@
+        return;
+      }
+      const memberName = sourceCode.getText(methodNode.key);
+      const className = methodNode.parent.parent.id?.name ?? 'class';
+
+      // Check in comparison to the base types.
+      for (const {
+        baseMemberType,
+        baseType,
+        heritageToken,
+      } of util.getBaseTypesOfClassMember(checker, methodTsNode)) {
+        const baseName = baseType.getSymbol()?.name ?? 'base';
+        if (isVoidReturningFunctionType(baseMemberType)) {
+          const msgId: ErrorPlaceId =
+            heritageToken === ts.SyntaxKind.ExtendsKeyword
+              ? 'ExtMember'
+              : 'ImplMember';
+          reportIfNonVoidFunction(methodNode.value, msgId, {
+            baseName,
+            className,
+            memberName,
+          });
+          return; // Report at most one error.
+        }
+      }
+    }
+
+    /**
+     * Reports an error if the provided node is not allowed in a void function context.
+     */
+    function reportIfNonVoidFunction(
+      funcNode: TSESTree.Expression,
+      msgId: ErrorPlaceId,
+      data?: Record<string, string>,
+    ): void {
+      const allowedReturnType =
+        ts.TypeFlags.Void |
+        ts.TypeFlags.Never |
+        (options.allowReturnUndefined ? ts.TypeFlags.Undefined : 0) |
+        (options.allowReturnNull ? ts.TypeFlags.Null : 0) |
+        (options.allowReturnAny ? ts.TypeFlags.Any : 0);
+
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(funcNode);
+      const actualType = checker.getApparentType(
+        checker.getTypeAtLocation(tsNode),
+      );
+
+      if (
+        tsutils
+          .getCallSignaturesOfType(actualType)
+          .map(signature => signature.getReturnType())
+          .flatMap(returnType => tsutils.unionTypeParts(returnType))
+          .every(type => tsutils.isTypeFlagSet(type, allowedReturnType))
+      ) {
+        // The function is already void.
+        return;
+      }
+
+      if (
+        funcNode.type !== AST_NODE_TYPES.ArrowFunctionExpression &&
+        funcNode.type !== AST_NODE_TYPES.FunctionExpression
+      ) {
+        // The provided function is not a function literal.
+        // Report a generic error.
+        return context.report({
+          node: funcNode,
+          messageId: `nonVoidFuncIn${msgId}`,
+          data,
+        });
+      }
+
+      // The provided function is a function literal.
+
+      if (funcNode.generator) {
+        // The provided function is a generator function.
+        // Generator functions are not allowed.
+        return context.report({
+          loc: util.getFunctionHeadLoc(funcNode, sourceCode),
+          messageId: `nonVoidFuncIn${msgId}`,
+          data,
+        });
+      }
+
+      if (funcNode.async) {
+        // The provided function is an async function.
+
+        if (!options.allowReturnPromiseIfTryCatch) {
+          // Async functions aren't allowed.
+          // TODO: Suggest wrapping its body in an async IIFE.
```

Comment:
[Docs] Could you remove these TODOs actually? We haven't decided what the right next step would be. I'm personally loathe to suggest async IIFEs; IMO they're 99% of the time an anti-pattern that results in dangerous async patterns.

Reply:
I agree that AIIFEs suck, but that's the only way (that I know) to make this rule (or no-misused-promises) happy. That's why I added that suggestion and ignoreAsyncInTryCatch.

Reply:
Removed